tests/torch_api/test_broadcast_state.py

```diff
@@ -0,0 +1,182 @@
+
+
+def run_bagua_broad(rank, nprocs, bagua_params, envs, opt_class, opt_hyper_param):
+    _init_bagua_env(rank, envs)
+
+    bagua_model, bagua_optimizer = create_model_and_optimizer(
+        opt_class, opt_hyper_param
+    )
+
+    from bagua.torch_api.algorithms import gradient_allreduce
+    algorithm = gradient_allreduce.GradientAllReduceAlgorithm()
+
+    bagua_model = bagua_model.with_bagua([bagua_optimizer], algorithm)
+    try:
+        bagua_model = bagua_model.with_bagua([bagua_optimizer], algorithm)
+    except Exception:
+        time.sleep(0.1)
+
+    model_params = [
+        (k, v.clone().detach().cpu().numpy())
+        for k, v in sorted(bagua_model.state_dict().items())
+    ]
+    optimizer_params = get_optimizer_param_values(bagua_optimizer)
+
+    # Put "model_params" in dimension 1, while "optimizer_params" in dimension 2.
+    bagua_params[rank][0].extend(model_params)
+    bagua_params[rank][1].extend(optimizer_params)
+
+
+class Test_Broadcast_Module(unittest.TestCase):
+    def test_broadcast_module(self):
+        nprocs = torch.cuda.device_count()
+        optimizers = [
+            (subclass.__name__, subclass)
+            for subclass in torch.optim.Optimizer.__subclasses__()
+            if subclass.__module__.startswith("torch.optim") and subclass != torch.optim.LBFGS and subclass != torch.optim.SparseAdam
+        ]
+
+        optimizer_hyper_param = [
+            dict(lr=0.2, momentum=0.9, weight_decay=0.1, centered=True),
+            dict(lr=0.2),
+        ]
+
+        for (opt_name, opt_class), opt_hyper_param in itertools.product(
+            optimizers, optimizer_hyper_param
+        ):
+            env = {
+                "WORLD_SIZE": str(nprocs),
+                "LOCAL_WORLD_SIZE": str(nprocs),
+                "MASTER_ADDR": "127.0.0.1",
+                "MASTER_PORT": str(find_free_port(8000, 8100)),
+                "BAGUA_SERVICE_PORT": str(find_free_port(9000, 9100)),
+            }
+            with Manager() as manager:
+                # For each rank, set a two dimensional list. One is used to save model_params,
+                # while the second save optimizer_params.
+                bagua_params = manager.list(
+                    [[manager.list() for _ in range(2)] for _ in range(nprocs)]
+                )
+                mp = multiprocessing.get_context("spawn")
+                processes = []
+                for i in range(nprocs):
+                    p = mp.Process(
+                        target=run_bagua_broad,
+                        args=(
+                            i,
+                            nprocs,
+                            bagua_params,
+                            env,
+                            opt_class,
+                            opt_hyper_param,
+                        ),
+                    )
+                    p.start()
+                    processes.append(p)
+                for p in processes:
+                    p.join(timeout=60)
+
```

Comment:
We can add a statement `p.exitcode == 0` here to check whether the process terminates normally.

Reply:
ok, i will add it and confirm.